### PR TITLE
ORS Sample Draft

### PR DESCRIPTION
**Pull request by t-jadang-msft**
ORS Sample program that uploads 5 blobs and tracks the progress of the replication. Once replication is complete, it updates one blob and verifies the replication. Finally, it moves all blobs to archive tier.

### CODE REVIEW

File: ORSSample/main.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+    print("Archiving...")
+    archive_blobs(source_blob_container_client, blobs)
+    # archive_batch(source_blob_container_client, blobs)
+    print("Archived")
+
+
+# Upload multiple blobs
+def upload_blobs(source_blob_service_client, container_name, blobs):
+    for x in blobs:
```
Comment (t-cezhan-msft) [inline ORSSample/main.py]: Maybe worth having a more descriptive variable name? So maybe blob_name instead of x

File: ORSSample/main.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+# Tracks progress of replicating multiple blobs
+def track_progress(source_blob_service_client, container_name, blobs):
+    completed = 0
+    length = len(blobs)
+    incomplete_events = blobs.copy()
+    completed_events = []
+
+    while completed < length:
+        for x in incomplete_events:
```
Comment (t-cezhan-msft) [inline ORSSample/main.py]: Same here, maybe incomplete_event instead of x

File: ORSSample/main.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    print(download_dest.read())
+    download_dest.close()
+
+    os.remove("download_source.txt")
+    os.remove("download_dest.txt")
+
+
+# Archive blobs individually
+def archive_blobs(container_client, blobs):
+    for x in blobs:
```
Comment (t-cezhan-msft) [inline ORSSample/main.py]: Same here as well

File: ORSSample/main.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+source_connect_str = SOURCE_CONNECTION_STRING
+dest_connect_str = DESTINATION_CONNECTION_STRING
+source_container_name = SOURCE_CONTAINER_NAME
+dest_container_name = DESTINATION_CONTAINER_NAME
+blobs = BLOB_NAMES
+
+
+def main():
+
```
Comment (t-cezhan-msft) [inline ORSSample/main.py]: Remove whitespace

File: ORSSample/main.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    source_blob_service_client = BlobServiceClient.from_connection_string(source_connect_str)
+    source_blob_container_client = source_blob_service_client.get_container_client(source_container_name)
+    source_blob_client = source_blob_service_client.get_blob_client(container=source_container_name, blob=blobs[0])
+
+    print("Uploading source blobs...")
+    upload_blobs(source_blob_service_client, source_container_name, blobs)
+
+    print("Replicating...")
+    # check_completion(source_blob_client)
```
Comment (t-cezhan-msft) [inline ORSSample/main.py]: Uncomment/remove commented out code, or explain why it is commented

File: ORSSample/main.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
+import os
```
Comment (t-cezhan-msft) [inline ORSSample/main.py:1]: Remember to add a README

File: ORSSample/main.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    source_blob_client.upload_blob("Lorem Ipsum", overwrite=True)
+
+    print("Replicating...")
+    check_completion(source_blob_client)
+    print("Replication Complete")
+    print_contents(source_blob_client, dest_blob_client)
+
+    print("Archiving...")
+    archive_blobs(source_blob_container_client, blobs)
+    # archive_batch(source_blob_container_client, blobs)
```
Comment (t-cezhan-msft) [inline ORSSample/main.py]: Uncomment/remove commented out code, or explain why it is commented

File: ORSSample/main.py [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+        for x in incomplete_events:
+            blob_client = source_blob_service_client.get_blob_client(container=container_name, blob=x)
+            props = blob_client.get_blob_properties()
+
+            for replication_policy in props.object_replication_source_properties:
+                for rule in replication_policy.rules:
+                    if rule.status == "complete":
+                        completed += 1
+                        completed_events.append(x)
+                        incomplete_events.remove(x)
```
Comment (t-cezhan-msft) [inline ORSSample/main.py]: It's probably best if you don't modify what you're iterating over. So instead, maybe do "for incomplete_event in incomplete_events[:]:" so you iterate over a copy of the original list

File: ORSSample/main.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+
+# Archive blobs individually
+def archive_blobs(container_client, blobs):
+    for blob_name in blobs:
+        blob_client = container_client.get_blob_client(blob_name)
+        blob_client.set_standard_blob_tier("Archive")
+
+
+# Archive blobs using a batch
+def archive_batch(container_client, blobs):
```
Comment (xiafu-msft) [inline ORSSample/main.py:126]: >archive_batch [](start = 4, length = 13)

So you leave it there on purpose so that after service gets fixed we can use it?

Comment (t-jadang-msft) [reply ORSSample/main.py:126]: Yes, I included the archive batch and delete batch functions for when the service is fixed

File: ORSSample/main.py [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+    download_source = open("download_source.txt", "r")
+    print(download_source.read())
+    download_source.close()
+
+    print("Destination Blob Contents:")
+    download_dest = open("download_dest.txt", "wb")
+    download_dest.write(dest_blob_client.download_blob().readall())
+    download_dest = open("download_dest.txt", "r")
+    print(download_dest.read())
+    download_dest.close()
```
Comment (xiafu-msft) [inline ORSSample/main.py]: Why do we need to download into a file?
print(dest_blob_client.download_blob().readall()) can do the same thing?

**Review by xiafu-msft — APPROVED** [review]

**Pull request merged**